File: backend/app/payment_helpers/northernMichigan/traverseCity/tent/payTimberRidgeTent.py
from datetime import datetime
from playwright.sync_api import sync_playwright
import time


def pay_timberRidgeTent(start_date, end_date, num_adults, num_kids, payment_info):
    # Convert dates to the required format (DD-MM-YY)
    start_date_formatted = datetime.strptime(start_date, "%m/%d/%y").strftime("%d-%m-%y")
    end_date_formatted = datetime.strptime(end_date, "%m/%d/%y").strftime("%d-%m-%y")

    response_data = {
        "base_price": 0,
        "tax": 0,
        "total": 0,
        "payment_successful": False
    }

    url = f"https://bookingsus.newbook.cloud/timberridgeresort/index.php?available_from={start_date_formatted}&available_to={end_date_formatted}&adults={num_adults}&kids={num_kids}&equipment_type=3&equipment_length=20"

    with sync_playwright() as p:
        browser = p.chromium.launch(headless=True)
        context = browser.new_context()
        page = context.new_page()
        page.goto(url)
        time.sleep(1)

        try:
            page.wait_for_selector('#category-list')

            first_option = page.query_selector('.newbook_online_category_box')

            if first_option:
                first_option.wait_for_selector('.newbook_online_from_price_text', timeout=30000)
                price_element = first_option.query_selector('.newbook_online_from_price_text')

                if price_element:
                    price = price_element.inner_text()
                    book_button = first_option.query_selector('button.button.newbook_responsive_button.full_width_uppercase')
                    if book_button:
                        book_button.click()
                        time.sleep(2)

                        main_container = page.locator('#newbook_content')
                        main_container.wait_for(state='visible', timeout=5000)
                        
                        # Click confirm site selection button
                        confirm_button = main_container.locator('#generic_button_109')
                        confirm_button.wait_for(state='visible', timeout=5000)
                        confirm_button.click()
                        time.sleep(1)

                        # Click checkout button
                        # footer_container = page.locator('.newbook_online_footer')
                        # footer_container.wait_for(state='visible', timeout=5000)
                        # checkout_button = footer_container.locator('#generic_button_2')
                        # checkout_button.wait_for(state='visible', timeout=5000)
                        # checkout_button.click()
                        # time.sleep(1)

                        # fill out payment info
                        main_container = page.locator('#newbook_content')
                        main_container.wait_for(state='visible', timeout=5000)

                        newbook_row = main_container.locator('.newbook-row').nth(1)

                        first_name_input = newbook_row.locator('#firstname')
                        first_name_input.wait_for(state='visible', timeout=5000)
                        first_name_input.fill(payment_info['first_name'])

                        last_name_input = newbook_row.locator('#lastname')
                        last_name_input.wait_for(state='visible', timeout=5000)
                        last_name_input.fill(payment_info['last_name'])

                        email_input = newbook_row.locator('#contact_details_email')
                        email_input.wait_for(state='visible', timeout=5000)
                        email_input.fill(payment_info['email'])

                        address_input = newbook_row.locator('#google_address_lookup')
                        address_input.wait_for(state='visible', timeout=5000)
                        address_input.fill(payment_info['street_address'])

                        city_input = newbook_row.locator('#city')
                        city_input.wait_for(state='visible', timeout=5000)
                        city_input.fill(payment_info['city'])

                        zip_code_input = newbook_row.locator('#postcode')
                        zip_code_input.wait_for(state='visible', timeout=5000)
                        zip_code_input.fill(payment_info['zip_code'])

                        newbook_row = main_container.locator('#newbook_payment_method_options')
                        newbook_row.wait_for(state='visible', timeout=5000)

                        card_number_input = newbook_row.locator('#card_number')
                        card_number_input.wait_for(state='visible', timeout=5000)
                        card_number_input.fill(payment_info['card_number'])

                        card_expiry_input = newbook_row.locator('#card_expiry')
                        card_expiry_input.wait_for(state='visible', timeout=5000)
                        card_expiry_input.fill(payment_info['expiry_date'])

                        card_cvv_input = newbook_row.locator('#card_cvv')
                        card_cvv_input.wait_for(state='visible', timeout=5000)
                        card_cvv_input.fill(payment_info['cvc'])

                        card_name_input = newbook_row.locator('#card_name')
                        card_name_input.wait_for(state='visible', timeout=5000)
                        card_name_input.fill(payment_info['cardholder_name'])

                        agree_checkbox = main_container.locator('.newbook_checkout_checkbox').nth(2)
                        agree_checkbox.wait_for(state='visible', timeout=5000)
                        # Get the label element and click on a safe area that's not a hyperlink
                        label = agree_checkbox.locator('label[for="agree"]')
                        # Click on the beginning of the label text which contains "I/We have read"
                        label.click(position={"x": 10, "y": 10})

                        # Get the total price
                        total_element = page.locator('.sidebar_cart_total')
                        total_element.wait_for(state='visible', timeout=5000)
                        total_price = total_element.inner_text()
                        total_price = float(total_price.replace('$', '').strip())    
                        response_data["total"] = total_price

                        place_booking_button = main_container.locator('#place_booking_full')
                        place_booking_button.wait_for(state='visible', timeout=5000)
                        # uncomment to pay
                        # place_booking_button.click()
                        # time.sleep(2)

                        response_data["base_price"] = total_price - 3
                        response_data["tax"] = 3
                        response_data["payment_successful"] = True
                        return response_data 
                    else:
                        print("Book button not found. This should not happen. Investigate further.")
                        return False
                else:
                    print("price_element not found. This should not happen. Investigate further.")
                    return False
            else:
                print("first_option not found. This should not happen. Investigate further.")
                return False

        except Exception as e:
            print(f"An error occurred: {e}")
            return False

        finally:
            browser.close()

def main():
    payment_info = {
        "first_name": "Lebron",
        "last_name": "James",
        "email": "lebronjames@gmail.com",
        "phone_number": "3134321234",
        "street_address": "1234 Rocky Rd",
        "city": "San Francisco",
        "state": "CA",
        "zip_code": "45445",
        "country": "USA",
        "cardholder_name": "Lebron James",
        "card_number": "2342943844322224",
        "card_type": "Visa",
        "expiry_date": "01/30",
        "cvc": "1234"
    }
    timberRidgeData = pay_timberRidgeTent('06/04/25', '06/11/25', 3, 1, payment_info)
    print(timberRidgeData)

if __name__ == '__main__':
    main()

File: backend/app/scrapers-local-testing/pay_timberRidgeTent.js
const puppeteer = require('puppeteer');

async function payTimberRidgeTent(startDate, endDate, numAdults, numKids, paymentInfo) {
  console.log('Starting Timber Ridge Tent booking process...');
  const responseData = {
    base_price: 0,
    tax: 0,
    total: 0,
    payment_successful: false,
  };

  // Format dates
  const startDateFormatted = new Date(startDate).toISOString().split('T')[0];
  const endDateFormatted = new Date(endDate).toISOString().split('T')[0];
  console.log(`Formatted dates: ${startDateFormatted} to ${endDateFormatted}`);
  
  const url = `https://bookingsus.newbook.cloud/timberridgeresort/index.php?available_from=${startDateFormatted}&available_to=${endDateFormatted}&adults=${numAdults}&kids=${numKids}&equipment_type=3&equipment_length=20`;
  console.log(`Navigating to URL: ${url}`);
  
  try {
    console.log('Launching browser...');
    const browser = await puppeteer.launch({
      headless: false,
      args: ["--no-sandbox", "--disable-setuid-sandbox"]
    });

    const page = await browser.newPage();
    await page.setExtraHTTPHeaders({
      'user-agent': 'Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/125.0.0.0 Safari/537.36',
      'sec-ch-ua': '"Chromium";v="125", "Not.A/Brand";v="24"',
      'sec-ch-ua-mobile': '?0',
      'sec-ch-ua-platform': '"macOS"',
    });

    console.log('Waiting for page load...');
    await page.goto(url, { waitUntil: 'networkidle2' });
    await new Promise(resolve => setTimeout(resolve, 1000));

    console.log('Waiting for category list...');
    await page.waitForSelector('#category-list');
    const firstOption = await page.$('.newbook_online_category_box');

    if (firstOption) {
        console.log('Found first booking option');
        await page.waitForSelector('.newbook_online_from_price_text', { timeout: 30000 });
        const priceElement = await firstOption.$('.newbook_online_from_price_text');

        if (priceElement) {
            const price = await priceElement.evaluate(el => el.innerText);
            console.log(`Initial price found: ${price}`);
            const bookButton = await firstOption.$('button.button.newbook_responsive_button.full_width_uppercase');
            
            if (bookButton) {
                console.log('Clicking book button...');
                await bookButton.click();
                await new Promise(resolve => setTimeout(resolve, 2000));

                console.log('Confirming site selection...');
                await page.waitForNavigation({ waitUntil: 'networkidle0' });
                await page.waitForSelector('#newbook_content', { visible: true, timeout: 5000 });
                try {
                    // Wait for any button with text "Confirm site selection"
                    await page.waitForFunction(
                        () => [...document.querySelectorAll('button')].find(button => 
                            button.textContent.trim().includes('Confirm site selection')
                        ),
                        { timeout: 10000 }
                    );
                    
                    // Click the button
                    await page.evaluate(() => {
                        const button = [...document.querySelectorAll('button')].find(button => 
                            button.textContent.trim().includes('Confirm site selection')
                        );
                        if (button) button.click();
                    });
                } catch (error) {
                    console.log('Error in confirm site selection: ', error);
                    console.log('Could not find Confirm site selection button, printing all buttons:');
                    const buttons = await page.$$eval('button', buttons => buttons.map(btn => ({
                        id: btn.id,
                        text: btn.textContent,
                        class: btn.className
                    })));
                    console.log(JSON.stringify(buttons, null, 2));
                    throw error;
                }
                await new Promise(resolve => setTimeout(resolve, 3000));

                console.log('Filling payment information...');
                await page.waitForNetworkIdle();
                await page.waitForSelector('#firstname', { visible: true, timeout: 5000 });
                await page.type('#firstname', paymentInfo.first_name);
                await page.type('#lastname', paymentInfo.last_name);
                await page.type('#contact_details_email', paymentInfo.email);
                await page.type('#google_address_lookup', paymentInfo.street_address);
                await page.type('#city', paymentInfo.city);
                await page.type('#postcode', paymentInfo.zip_code);

                console.log('Filling card details...');
                await page.waitForSelector('#card_number', { visible: true, timeout: 5000 });
                await page.type('#card_number', paymentInfo.card_number);
                await page.type('#card_expiry', paymentInfo.expiry_date);
                await page.type('#card_cvv', paymentInfo.cvc);
                await page.type('#card_name', paymentInfo.cardholder_name);

                console.log('Clicking agree checkbox...');
                const agreeCheckbox = await page.$$('.newbook_checkout_checkbox');
                await agreeCheckbox[2].click({ position: { x: 10, y: 10 } });

                console.log('Getting final price...');
                await page.waitForSelector('#place_booking_full', { visible: true, timeout: 5000 });
                const totalPrice = await page.$eval('#place_booking_full span', el => {
                    const priceMatch = el.innerText.match(/\$(\d+\.\d+)/);
                    return priceMatch ? parseFloat(priceMatch[1]) : 0;
                });
                console.log(`Total price found: $${totalPrice}`);

                responseData.total = totalPrice;
                responseData.base_price = totalPrice - 3;
                responseData.tax = 3;
                responseData.payment_successful = true;

                // Uncomment to actually place booking
                // console.log('Placing final booking...');
                // await page.click('#place_booking_full');
                // await new Promise(resolve => setTimeout(resolve, 2000));
            }
        }
    }

    console.log("Booking process completed. Response data: ", responseData);
    await browser.close();
    return responseData;

  } catch (error) {
    console.error(`Error during payment process: ${error.message}`);
    console.error(`Error stack: ${error.stack}`);
    return responseData;
  }
}

if (require.main === module) {
    (async () => {
        const paymentInfo = {
            first_name: "Lebron",
            last_name: "James",
            email: "lebronjames@gmail.com",
            phone_number: "3134321234",
            street_address: "1234 Rocky Rd",
            city: "San Francisco",
            state: "CA",
            zip_code: "45445",
            country: "USA",
            cardholder_name: "Lebron James",
            card_number: "2342943844322224",
            card_type: "Visa",
            expiry_date: "01/30",
            cvc: "1234"
        }
        const result = await payTimberRidgeTent("06/04/25", "06/06/25", 2, 1, paymentInfo);
        console.log("Scrape result:", result);
    })();
}

module.exports = { payTimberRidgeTent };